### PR TITLE
Add yvaults as an optional dependency to scope and re-use functions for kToken price calcs

### PR DESCRIPTION
**Pull request by elliotkennedy**
### Overview

Use yvaults code to calculate kToken prices.

Adds support for Raydium kToken prices.

### Notes

- Excludes uncompounded reward tokens from the price as they can likely be more easily manipulated than the pair tokens
- Takes the oldest component price updated slot and timestamp (excluding rewards)
- Increases the CU by `35000` (`50000` -> `85000`) - this could be optimised by splitting the yvaults functions to exclude reward tokens altogether.

#### localnet

```sh
id=108, entry='kSOL-bSOL_Orca/USD', price='0.9089980', price_type='KToken', age=-220614794, age_c=-220614794, max_age=250
id=109, entry='kMNDE-MSOL_Orca/USD', price='1.3945030', price_type='KToken', age=-220614693, age_c=-220614693, max_age=250
id=110, entry='kSTSOL-USDC_Orca/USD', price='1.0624320', price_type='KToken', age=-220614789, age_c=-220614789, max_age=250
id=111, entry='kUSDH-USDT_Orca/USD', price='1.0335760', price_type='KToken', age=-220614824, age_c=-220614824, max_age=250
id=112, entry='kSOL-JITOSOL_Orca/USD', price='0.0000530', price_type='KToken', age=-220614794, age_c=-220614794, max_age=250
id=113, entry='kbSOL-MSOL_Orca/USD', price='0.0000480', price_type='KToken', age=-220614693, age_c=-220614693, max_age=250
id=114, entry='kMSOL-JITOSOL_Orca/USD', price='1.2568160', price_type='KToken', age=-220614693, age_c=-220614693, max_age=250
id=115, entry='kSOL-USDC_Orca/USD', price='1.0232520', price_type='KToken', age=-220614794, age_c=-220614794, max_age=250
id=116, entry='kJITOSOL-USDC_Orca/USD', price='0.0411940', price_type='KToken', age=-220614794, age_c=-220614794, max_age=250
```

#### mainnet

```sh
2023-09-30T01:46:00.58707023Z stdout F id=108, entry='kSOL-bSOL_Orca/USD', price='0.9086510', price_type='KToken', age=237, age_c=237, max_age=250
2023-09-30T01:46:00.587130022Z stdout F id=109, entry='kMNDE-MSOL_Orca/USD', price='1.3938180', price_type='KToken', age=199, age_c=199, max_age=250
2023-09-30T01:46:00.587260539Z stdout F id=110, entry='kSTSOL-USDC_Orca/USD', price='1.0623260', price_type='KToken', age=220, age_c=220, max_age=250
2023-09-30T01:46:00.587465309Z stdout F id=111, entry='kUSDH-USDT_Orca/USD', price='1.0335660', price_type='KToken', age=146, age_c=146, max_age=250
2023-09-30T01:46:00.587555633Z stdout F id=112, entry='kSOL-JITOSOL_Orca/USD', price='0.0000530', price_type='KToken', age=237, age_c=237, max_age=250
2023-09-30T01:46:00.587609217Z stdout F id=113, entry='kbSOL-MSOL_Orca/USD', price='0.0000480', price_type='KToken', age=237, age_c=237, max_age=250
2023-09-30T01:46:00.587761619Z stdout F id=114, entry='kMSOL-JITOSOL_Orca/USD', price='1.2557860', price_type='KToken', age=237, age_c=237, max_age=250
2023-09-30T01:46:00.587838215Z stdout F id=115, entry='kSOL-USDC_Orca/USD', price='1.0231490', price_type='KToken', age=167, age_c=167, max_age=250
2023-09-30T01:46:00.588002037Z stdout F id=116, entry='kJITOSOL-USDC_Orca/USD', price='0.0411870', price_type='KToken', age=194, age_c=194, max_age=250
```

### CODE REVIEW

File: programs/scope/src/oracles/ktokens/kamino.rs [hunk 1]
```diff
@@ -1,1014 +0,0 @@
...
-            let px_x = calc_price_from_sqrt_price(x, decimals_a, decimals_b);
-            let px_y = calc_price_from_sqrt_price(y, decimals_a, decimals_b);
-
-            let diff = (px_x - px_y).abs();
-            println!("x: {x}, y: {y} diff: {diff}");
-        }
-    }
-
-    #[test]
-    fn scope_prices_to_sqrt_prices() {
```
Comment (y2kappa) [inline programs/scope/src/oracles/ktokens/kamino.rs:841]: what is happening to these tests? they are not in yvaults also

File: programs/scope/src/oracles/ktokens.rs [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+    let token_prices = kamino::utils::scope::get_prices_from_data(
+        scope_prices_ref.deref(),
+        &collateral_infos_ref.infos,
+        &strategy_account_ref,
+        Some(clmm.as_ref()),
+        clock.slot,
+    )?;
+
+    let holdings = common::holdings(
```
Comment (y2kappa) [inline programs/scope/src/oracles/ktokens.rs:119]: This is not a 1-2-1 translation of the previous implementation, by abstracting it away into what yvaults does, you lost this piece of logic:

https://github.com/hubbleprotocol/scope/pull/76/files#diff-bb54d2d78b9726cfd5f19b5f68a039706362ca1caa219eb13fa90f51a1bd76a0L118

File: programs/scope/src/oracles/ktokens.rs [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+                protocol_position: None,
+                lower_tick_array: None,
+                upper_tick_array: None,
+            })
+        }
+    };
+    Ok(clmm)
+}
+
+fn get_component_px_last_update(
```
Comment (oeble) [inline programs/scope/src/oracles/ktokens.rs:196]: Can you add the design choices in a mod doc. Like here, the fact that only token a/b is reflected in the timestamp and not the rewards.

File: programs/scope/src/oracles/ktokens.rs [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    let holdings_value = holdings
+        .total_sum
+        .checked_sub(holdings.rewards_usd)
+        .ok_or(ScopeError::IntegerOverflow)?;
+
+    let token_price = get_price_per_full_share_impl(
+        &holdings_value,
+        strategy_account_ref.shares_issued,
+        strategy_account_ref.shares_mint_decimals,
+    )?;
```
Comment (oeble) [inline programs/scope/src/oracles/ktokens.rs:136]: For the follow-up, reimplementing `get_price_per_full_share_impl` to output with dynamic or way more decimals (like 18 on `u128`) is probably not too hard. Then, we can have a heuristic `match` to select the final number of decimals based on the value range.

File: programs/scope/src/oracles/ktokens.rs (hunk 2 above)
Comment (elliotkennedy) [reply programs/scope/src/oracles/ktokens.rs]: Resolved here as discussed - https://github.com/hubbleprotocol/scope/pull/158/commits/cb08a3b3f1e84c51b25c39949cd1d9eb7577e374

File: programs/scope/src/oracles/ktokens/kamino.rs (hunk 1 above)
Comment (elliotkennedy) [reply programs/scope/src/oracles/ktokens/kamino.rs:841]: Good spot - resolved here as discussed - https://github.com/hubbleprotocol/scope/pull/158/commits/cb08a3b3f1e84c51b25c39949cd1d9eb7577e374

File: programs/scope/src/oracles/ktokens.rs (hunk 4 above)
Comment (elliotkennedy) [reply programs/scope/src/oracles/ktokens.rs]: Sounds good, will make a note

File: programs/scope/Cargo.toml [hunk 5]
```diff
@@ -17,13 +17,14 @@ debug = []
 
 # If none of the following is set, one will be selected based on env $CLUSTER variable
 # If $CLUSTER is not set either, default will be mainnet
+default = ["yvaults"]
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply programs/scope/Cargo.toml:20]: Fair enough - added

File: programs/scope/Cargo.toml [hunk 6]
```diff
@@ -34,8 +36,12 @@ spl-stake-pool = { version = "0.6.3", features = ["no-entrypoint"] }
 arrayref = "0.3.6"
 decimal-wad = "0.1.7"
 rust_decimal = "1.18.0"
-whirlpool = { git = "https://github.com/orca-so/whirlpools", features = ["no-entrypoint", "cpi"]}
 strum = { git = "https://github.com/hubbleprotocol/strum", features = ["derive"], branch = "checked_arithmetics" }
+# Comment out the line below if you do not have access to the yvaults repo
+yvaults = { git = "ssh://git@github.com/hubbleprotocol/yvaults.git", features = ["no-entrypoint", "cpi", "mainnet"], optional = true, branch = "elliot/refactor-px-functions" }
```
Comment (elliotkennedy) [inline programs/scope/Cargo.toml:41]: Changing back to `master` once https://github.com/hubbleprotocol/yvaults/pull/553 merged

**Comment by elliotkennedy** [comment]
Re-ran the clone mainnet accounts test :heavy_check_mark:

**Review by oeble — APPROVED** [review]

File: programs/scope/src/oracles/mod.rs [hunk 7]
```diff
@@ -51,7 +52,7 @@ impl OracleType {
             OracleType::SwitchboardV2 => 30000,
             OracleType::CToken => 130000,
             OracleType::SplStake => 20000,
-            OracleType::KToken => 50000,
+            OracleType::KToken => 90000,
```
[the remark this answers is not in the document]
Comment (elliotkennedy) [reply programs/scope/src/oracles/mod.rs:55]: It's the prices and extra rewards logic I think, can definitely be optimised if it's a problem.

**Pull request merged**